include/zephyr/sys/ref.h

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright (c) 2023 Michael Zimmermann
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_SYS_REF_H_
+#define ZEPHYR_INCLUDE_SYS_REF_H_
+
+#include <limits.h>
+#include <stdbool.h>
+#include <zephyr/sys/atomic.h>
+#include <zephyr/sys/__assert.h>
+#include <zephyr/toolchain.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @defgroup sys_ref_apis Reference count APIs
+ * @{
+ */
+
+/* atomics are usually `long`. If they are larger, that's fine because
+ * we'll just artificially limit them . If they were smaller, we'd be unable
+ * to check if an overflow happened. This wouldn't be necessary if the atomic
+ * API told us the max value of an atomic.
+ */
+BUILD_ASSERT(sizeof(atomic_val_t) >= sizeof(long));
+
+/**
+ * @brief An atomic reference count.
+ *
+ * The main use case is to share the same data between several users and free
+ * it when all references were released. This can also be used on SMP systems.
+ */
+struct sys_ref {
+	atomic_t refs;
```

Comment:
Minor: I feel like `count` would be a better name. The feature is called a "reference count" but you don't have "count" anywhere in the struct name, so put it into the internal counter variable name :)